dotnet add package cli support for cpm projects

pragnya17 opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug
Fixes: https://github.com/NuGet/Home/issues/11890
Regression? No.

## Description
This feature is based on the design spec: https://github.com/NuGet/Home/pull/11915. The goal was to allow the dotnet add package command to be executed for projects that were onboarded to CPM. Previously, the command threw a NU1008 error whenever packages were attempted to be added to projects onboarded to CPM. However, with this change users will now be able to add or update package references in projects onboarded to CPM. Directory.Packages.props files and project files are now modified accordingly whenever a package is added/updated.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - Functional test that checks if a package version is added to the props file correctly when there are no existing package versions in the props file already. It also checks that the package reference is added to the project file and that it does not have the package version metadata included in it.
  - Functional test that checks if a package version is added to the props file correctly when there already other existing package versions in the props file. It also checks that the package reference is added to the project file and that it does not have the package version metadata included in it.
  - Functional test that checks if a package version is updated with the correct package version attribute in the Directory.Packages.props file.

- **Documentation**
  - [x] Documentation PR or issue filled: https://github.com/NuGet/Home/issues/13480

pragnya17 commented on test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecOperationsTests.cs:
```diff
@@ -253,6 +253,102 @@ public void AddOrUpdateDependency_ToSpecificFrameworks_AddsNewDependency()
                 spec.TargetFrameworks[1].Dependencies[0].LibraryRange.VersionRange.MinVersion);
         }
 
+        [Fact]
```
All the modifications in this file are already in review based on @jeffkl  PR (https://github.com/NuGet/NuGet.Client/pull/4642) so I will remove these changes once his PR is merged into dev.

jeffkl replied:
I merged my changes so you can rebase now

pragnya17 commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecOperations.cs:
```diff
@@ -26,7 +26,7 @@ public static void AddOrUpdateDependency(PackageSpec spec, PackageDependency dep
 
             if (!existing.Any())
             {
```
All the modifications in this file are already in review based on @jeffkl  PR (https://github.com/NuGet/NuGet.Client/pull/4642) so I will remove these changes once his PR is merged into dev.

nkolev92 replied:
can you rebase that way the changeset is minimized and a bit easier to review?

pragnya17 replied:
Agreed, this has been resolved.

jeffkl commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -613,5 +614,215 @@ public async Task AddPkg_WhenPackageSourceMappingConfiguredCanotInstallsPackages
         }
+
+        [Fact]
+        public async Task AddPkgWithNoVersion_WhenProjectOnboardedToCPMAndNoPackagesExistInProps()
```
I would have this test create at least two package versions so that its testing that the latest is what's used.

pragnya17 replied:
That's a good point, I agree.

jeffkl commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -753,4 +753,8 @@ The search is a case-insensitive string comparison using the supplied value, whi
     <comment>0 - The command name. Ex. Push/Delete. 1 - list of server uris</comment>
   </data>
+  <data name="Info_AddPkgCPM" xml:space="preserve">
+    <value>PackageVersion and PackageReference for package '{0}' added to Directory.Packages.props file at '{1}' and the project file at '{2}' respectively.</value>
```
```suggestion
    <value>PackageReference for package '{0}' added to '{1}' and PackageVersion added to central package management file '{2}'.</value>
```

Just be sure to adjust the `string.Format()` args as well if you like this wording

pragnya17 replied:
This wording does seem more straightforward, I will go ahead and use that.

jeffkl commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:
```diff
@@ -152,27 +155,186 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
+            var packageVersion = AddVersionMetadata(libraryDependency, item);
+
+            Logger.LogInformation(string.Format(CultureInfo.CurrentCulture,
+            Strings.Info_AddPkgAdded,
```
nit: Fix indentation here, on line 284, and on line 305

jeffkl replied:
The indentation is still incorrect here, probably got reset by some other change in the editor.

jeffkl commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:
```diff
@@ -152,27 +155,186 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
+        /// </summary>
+        /// <param name="libraryDependency">Package Dependency of the package to be added.</param>
+        /// <param name="item">Item to add the metadata to.</param>
+        private void AddExtraMetadata(LibraryDependency libraryDependency, ProjectItemElement item)
```
nit: I would give this method a more specific name like `AddMetadataForLibraryDependency()` or something.  Also the same for `UpdateExtraMetadata()`

pragnya17 replied:
Makes sense, I updated both names to be more descriptive.

jeffkl commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:
```diff
@@ -152,27 +155,186 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
+        private ProjectRootElement GetDirectoryBuildPropsRootElement(Project project)
+        {
+            // Get the Directory.Packages.props path.
+            string directoryPackagesPropsPath = project.GetPropertyValue("DirectoryPackagesPropsPath");
```
Let's add a constant for the MSBuild property name `DirectoryPackagesPropsPath` as the string is referenced more than once in this file.

```c#
/// <summary>
/// The name of the MSBuild property that represents the path to the central package management file, usually Directory.Packages.props.
/// </summary>
private const string DirectoryPackagesPropsPathPropertyName = "DirectoryPackagesPropsPath";

pragnya17 replied:
This is a good point, I have added it as a constant.

jeffkl commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -613,5 +614,217 @@ public async Task AddPkg_WhenPackageSourceMappingConfiguredCanotInstallsPackages
+        }
+
+        [Fact]
+        public async Task AddPkgWithVersion_WhenProjectOnboardedToCPMAndPackageVersionExistsInProps()
```
Some people put effort into keeping their MSBuild project files orderly.  The code ensures that a new `<PackageVersion />` item is added to the first `<ItemGroup />` that contains a `<PackageVersion />` but the tests aren't verifying that.  Can you add a few that just test this logic?

Example Directory.Packages.props:
```xml
<Project>
  <PropertyGroup>
    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
  </PropertyGroup>
  <ItemGroup>
    <Content Include="SomeFile" />
  </ItemGroup>
  <ItemGroup>
    <!-- The test ensures that a new <PackageVersion /> item was added to this ItemGroup -->
    <PackageVersion Include="PackageA" Version="1.0.0" />
    <PackageVersion Include="PackageB" Version="1.0.0" />
  </ItemGroup>
</Project>
```

pragnya17 replied:
This is a great point that I had not considered, a test that reflects this has been added.

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:
```diff
@@ -152,27 +159,178 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
-                UpdatePackageReferenceItems(existingPackageReferences, libraryDependency);
+                // Get package version if it already exists in the props file. Returns null if there is no matching package version.
+                ProjectItem packageVersion = project.Items.LastOrDefault(i => i.ItemType == PACKAGE_VERSION_TYPE_TAG && i.EvaluatedInclude.Equals(libraryDependency.Name));
+                var versionCLIArgument = libraryDependency.LibraryRange.VersionRange.OriginalString;
```
The `libraryDependency` variable in this context is not aware if the version is passed from the command line or not.

There are 2 ways to identify that one is using [`PackageRefernceArgs.NoVersion`](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/PackageReferenceArgs.cs#L13) and other option is to check [`PackageReferenceArgs.PackageVersion`](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/PackageReferenceArgs.cs#L22) is null or empty.

pragnya17 replied:
This function has not been changed from the original implementation since it is referring to updating package reference items when a project is not onboarded to CPM. So I believe it does not need to be aware if the version is passed from the command line or not.

nkolev92 replied:
How do you differentiate between a version being passed on the commandline and a version selected as latest from the source.

pragnya17 replied:
After discussing with @jeffkl and @kartheekp-ms I added a parameter to all the `AddPackageReference` functions. This parameter is called `PackageVersion` and uses `PackageReferenceArgs.PackageVersion`. `PackageReferenceArgs.PackageVersion` is set based on if a version is passed as a CLI argument. This should resolve the issue.

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:
```diff
@@ -152,27 +159,178 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
             {
-                // If the package already has a reference then try to update the reference.
-                UpdatePackageReferenceItems(existingPackageReferences, libraryDependency);
+                // Get package version if it already exists in the props file. Returns null if there is no matching package version.
```
I think we should add `if (!existingPackageReferences.Any())` condition also because we are currently addressing first 4 scenarios described in https://github.com/NuGet/Home/pull/11915. In the existing implementation, code checks for existing `PackageReferences` for the same package id irrespective of package version. That's how we know that we should add `PackageReference` to the project file.

https://github.com/NuGet/NuGet.Client/blob/60b3c340843a04f056bb7b4c61f2400b19698ac1/src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs#L491-L504

pragnya17 replied:
In the first 4 scenarios that I am considering, the package reference does not exist. So in all of those scenarios a package reference must be added. Therefore I don't think it is necessary to check for existing package references, do you agree?

pragnya17 replied:
After discussing, we agreed to add this as it reflects the first 4 scenarios, all of which assume there are no existing package references.

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:
```diff
@@ -152,27 +159,178 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
+            var item = itemGroup.AddItem(PACKAGE_REFERENCE_TYPE_TAG, libraryDependency.Name);
+            var packageVersion = AddVersionMetadata(libraryDependency, item);
+
+            Logger.LogInformation(string.Format(CultureInfo.CurrentCulture, Strings.Info_AddPkgAdded, libraryDependency.Name, packageVersion, itemGroup.ContainingProject.FullPath));
```
I guess we should add this warning after updating the project file and Directory.Packages.props file. What do you think?

pragnya17 replied:
My current implementation does do that. This actually has not changed from the original implementation since this function is only called in non-CPM cases when the package reference must be updated. After the item is added to the Item Group and the version metadata is added in the .proj file, the information is logged. Does this align with what you were looking for?

pragnya17 replied:
To make it more clear, I logged information after all metadata is added.

pragnya17 committed:
Add package support for cpm projects

pragnya17 committed:
fix bug in test

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -753,4 +753,9 @@ The search is a case-insensitive string comparison using the supplied value, whi
   </data>
+  <data name="Info_AddPkgCPM" xml:space="preserve">
+    <value>PackageReference for package '{0}' added to '{1}' and PackageVersion added to central package management file '{2}'.</value>
+    <comment>[2:19 PM] Kartheek Penagamuri
```
Probably don't need `[2:19 PM] Kartheek Penagamuri` here :)

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:
```diff
@@ -152,27 +157,178 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
+        /// </summary>
+        /// <param name="libraryDependency">Package Dependency of the package to be added.</param>
+        /// <param name="item">Item to add the metadata to.</param>
+        private void AddExtraMetadataToProjectItemElement(LibraryDependency libraryDependency, ProjectItemElement item)
```
I think you have a bunch of great utility methods here. 
It'd be great to get some unit test. There's a lot of logic here that can probably be modified and no test would catch it.

pragnya17 replied:
That is a good point. Originally, I did consider adding unit tests however a lot of these utility methods were actually just created using code that already existed before I made modifications for CPM projects (I just put them into smaller methods)/they use the MSBuild API which is already tested. Considering that, I opted to write functional tests. 
I am totally open to the suggestion of unit tests though, so please let me know!

nkolev92 replied:
I think adding better coverage as unit tests is always ok. 

When you separate out a method even if it's from code that's already tested, it is something that can technically be misused. 

You also don't have to add tests for every little method, rather ones where it's convenient.

pragnya17 replied:
Unit tests have now been added.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:
```diff
@@ -152,27 +157,178 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
+                var versionCLIArgument = libraryDependency.LibraryRange.VersionRange.OriginalString;
+
+                //Add <PackageReference/> to the project file.
+                AddPackageReferenceIntoItemGroupCPM(project, itemGroup, libraryDependency);
```
I don't see any logic for version override.

pragnya17 replied:
The first 4 scenarios that I am implementing actually don't rely on VersionOverride at all so I didn't include it. I'm not sure if I should document that in the code to make it more clear maybe?

nkolev92 replied:
Ah, don't worry, that's my bad. 

The only additional thing that could be helpful is maybe copying the first 4 scenarios from the table in the issue description, but the link to the spec is already there, so it's easy to figure out.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -613,5 +613,268 @@ public async Task AddPkg_WhenPackageSourceMappingConfiguredCanotInstallsPackages
+            await SimpleTestPackageUtility.CreateFolderFeedV3Async(
+                    pathContext.PackageSource,
+                    PackageSaveMode.Defaultv3,
+                    packageX100);
```
```cs

await SimpleTestPackageUtility.CreateFolderFeedV3Async(
                    pathContext.PackageSource,
                    PackageSaveMode.Defaultv3,
                    packageX100,
                    packageX200);
````

You can combine them into 1.

nkolev92 replied:
Same applies to all tests.

pragnya17 replied:
Fixed, thanks for pointing that out!

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -613,5 +613,268 @@ public async Task AddPkg_WhenPackageSourceMappingConfiguredCanotInstallsPackages
         }
+
+        [Fact]
+        public async Task AddPkgWithNoVersion_WhenProjectOnboardedToCPMAndNoPackagesExistInProps()
```
naming: AddPkgWithNoVersion_WhenProjectOnboardedToCPMAndNoPackagesExistInProps

Consider reviewing:  https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md#unit-test-method-naming

`WhenProjectOnboardedToCPM` seems like a lot repeat in all test names. You're also missing the expected part in the name.

AddPkg_WithCPM_WhenPackageVersionDoesNotExist_XYZ what it may be.

pragnya17 replied:
Great point, I will go ahead and update the names! Thank you for pointing out the naming convention, I was not aware.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -613,5 +613,268 @@ public async Task AddPkg_WhenPackageSourceMappingConfiguredCanotInstallsPackages
+            var projectName = "projectA";
+            var projectA = XPlatTestUtils.CreateProject(projectName, pathContext, "net5.0");
+
+            const string version1 = "1.0.0";
```
nit: When you only use something in one spot, you don't need the property. 

Ideally tests are short and quick to read.

pragnya17 replied:
Noted and resolved.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -613,5 +613,268 @@ public async Task AddPkg_WhenPackageSourceMappingConfiguredCanotInstallsPackages
+            var result = _fixture.RunDotnet(projectADirectory, $"add {projectA.ProjectPath} package {packageX}", ignoreExitCode: true);
+
+            // Assert
+            Assert.True(result.Success, result.Output);
```
Great asserts!

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -613,5 +613,268 @@ public async Task AddPkg_WhenPackageSourceMappingConfiguredCanotInstallsPackages
+
+            // Assert
+            Assert.True(result.Success, result.Output);
+            Assert.Contains(@$"<ItemGroup>
```
We might have some utilities that clear the formatting and newlines, that might make these asserts easier. 
If you're not having trouble getting them correct, I wouldn't bother though.

pragnya17 replied:
Currently, all the tests pass with this formatting (I had to play around with it to get the correct formatting). However, if there is an easier way to fix the formatting/newlines I am definitely interested in learning about it!

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -613,5 +613,268 @@ public async Task AddPkg_WhenPackageSourceMappingConfiguredCanotInstallsPackages
+    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+  </PropertyGroup>
+  <ItemGroup>
+    <Content Include=""SomeFile"" />
```
You wouldn't normally have content in Directory.Packages.props. 

It'll only have 1 property and a bunch of PackageVersion items.

Consider removing it from all test cases. They don't really help you much.

pragnya17 replied:
The content was actually only added to the Props file in this one test because of @jeffkl 's comment here: https://github.com/NuGet/NuGet.Client/pull/4700#discussion_r908818570 . Hopefully that helps clarify things but please let me know if not.

nkolev92 replied:
I see. 

I guess it could only be done for some specific tests and not all, but I'll let you make a call on that.

pragnya17 committed:
Added package version parameter

nkolev92 reviewed (commented):
Great job so far.
A few suggestions to improve clarity and I think we might be missing some scenarios.

pragnya17 committed:
Modify test names to be more readable, remove comment in strings.resx

pragnya17 committed:
Change when info is logged

pragnya17 committed:
fix indentation and existing package references lines

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -613,5 +613,257 @@ public async Task AddPkg_WhenPackageSourceMappingConfiguredCanotInstallsPackages
+        }
+
+        [Fact]
+        public async Task AddPackageVersionToCorrectItemGroupInPropsFile()
```
This one doesn't have as great of a name as the rest in this file :)

pragnya17 replied:
Fixed the name.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -613,5 +613,257 @@ public async Task AddPkg_WhenPackageSourceMappingConfiguredCanotInstallsPackages
             Assert.Contains($"NU1100: Unable to resolve '{packageZ} (>= {version})' for 'net5.0'", result.AllOutput);
         }
+
+        [Fact]
```
The next 3 tests are great!
The names make it easy to understand what's being tested.

nkolev92 reviewed (commented):
This is looking great! 

Let's add some no-op and negative tests. 

In particular:

What if I specify a `--version` where the PackageVersion is already in the DBP. 
We should no-op in this case, or do whatever the `dotnet add package --version` command does on a standard PackageReference project. 

Think about the no-op version of every scenario you implemented.

kartheekp-ms committed:
[skip ci] add basic unit test for msbuild api

Commit:
Change name of test

Commit:
Add unit tests

pragnya17 commented:
> 

I assume that by DBP you mean the Directory.Build.props file? However, in the scenarios that I have implemented I assume that the DBP does not exist since it is out of the scope of my project. In this comment, Jeff also mentions that the Directory.Packages.props file is mandatory for onboarding to CPM: https://github.com/NuGet/Home/issues/11903#issuecomment-1161996051 .

I am happy to create a follow up work item to ensure that we fallback onto existing behavior for the scenario where the Directory.Packages.props file does not exist.

pragnya17 commented:
> This is looking great!
> 
> Let's add some no-op and negative tests.
> 
> In particular:
> 
> What if I specify a `--version` where the PackageVersion is already in the DBP. We should no-op in this case, or do whatever the `dotnet add package --version` command does on a standard PackageReference project.
> 
> Think about the no-op version of every scenario you implemented.

I assume that by DBP you mean the Directory.Build.props file? However, in the scenarios that I have implemented I assume that the DBP does not exist since it is out of the scope of my project. In this comment, Jeff also mentions that the Directory.Packages.props file is mandatory for onboarding to CPM: https://github.com/NuGet/Home/issues/11903#issuecomment-1161996051 .

I am happy to create a follow up work item to ensure that we fallback onto existing behavior for the scenario where the Directory.Packages.props file does not exist.

nkolev92 commented:
I meant to write if the version is already in DPP (Directory.Packages.props).

I don't really care about the scenarios where the project is not onboarded.

Commit:
Change name of integration test

Commit:
Change name to noop test

pragnya17 commented:
> I meant to write if the version is already in DPP (Directory.Packages.props).
> 
> I don't really care about the scenarios where the project is not onboarded.

Makes sense. I actually did already handle the scenario you were talking about (https://github.com/NuGet/NuGet.Client/pull/4700#pullrequestreview-1025590204), I just did not name it properly, so I have now updated the name of the test. For the first 4 scenarios that I implemented, after discussing with @kartheekp-ms and reviewing the spec again (https://github.com/NuGet/Home/blob/dev-kartheekp-ms-cpmaddpackagesupport-revised/proposed/2022/cpm-dotnet-cli-add-package-support.md), I don't believe there are any other no-op/fail scenarios that need to be included in the integration tests.

nkolev92 commented:
I'll take another look then!

nkolev92 reviewed (approved):
LGTM. 

Great job @pragnya17

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs:
```diff
@@ -57,7 +57,7 @@ public async Task<int> ExecuteCommand(PackageReferenceArgs packageReferenceArgs,
                 };
 
-                msBuild.AddPackageReference(packageReferenceArgs.ProjectPath, libraryDependency);
+                msBuild.AddPackageReference(packageReferenceArgs.ProjectPath, libraryDependency, packageReferenceArgs.PackageVersion);
```
I thought we changed this parameter to `bool` using `PackageReferenceArgs.NoVersion`. https://github.com/NuGet/NuGet.Client/blob/99e6dc73e6d23fce650327a21a5670a7039585db/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/PackageReferenceArgs.cs#L13

pragnya17 replied:
Must have gotten lost through some rebasing, it's now fixed.

kartheekp-ms commented on test/NuGet.Core.Tests/NuGet.CommandLine.Xplat.Tests/MSBuildAPIUtilityTests.cs:
```diff
@@ -0,0 +1,384 @@
+
+            // Assert
+            Assert.Contains(@$"<PackageReference Include=""X"" />", File.ReadAllText(Path.Combine(testDirectory, "projectA.csproj")));
+            Assert.DoesNotContain(@$"<Version = ""1.0.0"" />", File.ReadAllText(Path.Combine(testDirectory, "projectA.csproj")));
```
AFAIK, Version will not be added as a child element to the `PackageReference` item.  The format will be `<PackageReference Include="Newtonsoft.Json" Version="13.0.2-beta1" />`

```suggestion
Assert.DoesNotContain(@$"<PackageReference Include=""X"" Version = ""1.0.0"" />", File.ReadAllText(Path.Combine(testDirectory, "projectA.csproj")));
```
This suggestion applies to other tests that assert for `Version` attribute.

pragnya17 replied:
Agreed, will fix for all the tests. Thanks for the catch!

kartheekp-ms commented on test/NuGet.Core.Tests/NuGet.CommandLine.Xplat.Tests/MSBuildAPIUtilityTests.cs:
```diff
@@ -0,0 +1,384 @@
+
+            // Act
+            msObject.AddPackageReferenceIntoItemGroupCPM(project, itemGroup, libraryDependency);
+            project.Save();
```
Had an offline conversation with @pragnya17 about removing `project.Save()` method call but instead use `MSBuild Project` object model to validate the result. 

For context, `project.Save()` is invoked here to save the modification done by `AddPackageReferenceIntoItemGroupCPM` to the file system and then verify the behavior. Given that these are unit tests, I am wondering if we use the MSBuild API to validate the result instead of reading csproj or packages.props file.

pragnya17 replied:
I agree. However, there were some issues with debugging the unit test after making this modification so during an offline conversation, Kartheek and I agreed that he would be able to finish debugging/making these modifications to the unit tests.

Commit:
[skip ci] Use boolean to see if package version is passed in as CLI arg or not

Commit:
[skip ci] Attempt to modify unit test, still need to debug

kartheekp-ms commented on test/NuGet.Core.Tests/NuGet.CommandLine.Xplat.Tests/MSBuildAPIUtilityTests.cs:
```diff
@@ -0,0 +1,385 @@
+            project.Save();
+
+            // Assert
+            Assert.Contains(@$"<PackageReference Include=""X"" />", File.ReadAllText(Path.Combine(testDirectory, "projectA.csproj")));
+            Assert.DoesNotContain(@$"<Version = ""1.0.0"" />", File.ReadAllText(Path.Combine(testDirectory, "projectA.csproj")));
```
I think we can slightly improve the performance by reading text from the project only once.

pragnya17 replied:
Agreed, will fix that.

kartheekp-ms commented on test/NuGet.Core.Tests/NuGet.CommandLine.Xplat.Tests/MSBuildAPIUtilityTests.cs:
```diff
@@ -0,0 +1,385 @@
+
+            // Assert
+            Assert.Equal(projectContent, File.ReadAllText(Path.Combine(testDirectory, "projectA.csproj")));
+            Assert.Contains(@$"<PackageVersion Include=""X"" Version=""2.0.0"" />", File.ReadAllText(Path.Combine(testDirectory, "Directory.Packages.props")));
+            Assert.DoesNotContain(@$"<PackageVersion Include=""X"" Version=""1.0.0"" />", File.ReadAllText(Path.Combine(testDirectory, "Directory.Packages.props")));
```
I think we can slightly improve the performance by reading text from the props file only once.

pragnya17 replied:
Agreed, will fix that.

kartheekp-ms reviewed (approved):
Great job. I know CI is failing on other platforms. I will create a separate issue to track that work. Given that this work is getting merged into a feature branch, I will fix the CI scripts before creating a new pull request for merging these changes into the dev branch.

Commit:
Revert "[skip ci] Attempt to modify unit test, still need to debug"

This reverts commit 4e49d9360afb400af008066d5683c76ea266b30d.

Commit:
[skip ci] remove repetition when reading file from disk

kartheekp-ms committed:
updated tests

kartheekp-ms commented:
@pragnya17 - I have updated `MSBuildAPIUtilityTests.cs` to run only on Windows. All CI checks passed now for this PR. Please feel to merge this pull request. thanks

pragnya17 commented on test/NuGet.Core.Tests/NuGet.CommandLine.Xplat.Tests/MSBuildAPIUtilityTests.cs:
```diff
@@ -174,14 +177,15 @@ public void AddPackageReferenceIntoProjectFileWhenItemGroupDoesExist_Success()
-            Assert.DoesNotContain(@$"<Version = ""1.0.0"" />", File.ReadAllText(Path.Combine(testDirectory, "projectA.csproj")));
+            string updatedProjectFile = File.ReadAllText(Path.Combine(testDirectory, "projectA.csproj"));
+            Assert.Contains(@$"<PackageReference Include=""X"" />", updatedProjectFile);
+            Assert.DoesNotContain(@$"<Version = ""1.0.0"" />", updatedProjectFile);
```
This always return true, so I think it should be changed to $"PackageReference Includes =""X"" <Version = ""1.0.0"" />". This is along the lines of this comment: https://github.com/NuGet/NuGet.Client/pull/4700#discussion_r915421010
